RIOT-OS/RIOT
# sys/net/dhcpv6: Add stateless DHCPv6

### Contribution description

This PR adds stateless DHCP functionality to the DHCPv6 client. It allows a client to send information-requests to obtain additional configuration parameters from DHCP servers if the O bit in NDP Router Advertisements is set. If the M bit is set and the IA_NA module is included, the client switches to stateful mode instead. 

As the default configuration mode of the DHCPv6 client is now "inactive" by default, this PR might constitute a breaking change in certain respects. However, I added a respective function call to `client_simple_pd.c` which ensures that the client still behaves like before in the border router use case. I also updated the test under `tests/gnrc_dhcpv6_client`. Still, this aspect might require some discussion, including the way I implemented the state change itself.



### Testing procedure

I added an additional test under `tests/gnrc_dhcpv6_client_stateless`. The other DHCPv6 related tests should also still pass if you run them.

In the added test, the node first sends a NDP Router Solicition and in turn receives a Router Advertisement with a global prefix  for SLAAC and a set O bit, which implies that additional configuration information is available using DHCPv6. Using this information, the DHCPv6 client is reconfigured to use the `stateless` configuration mode and gets (re)started. Finally, it sends a DHCPv6 information-request which is answered by a reply message.

A number of assertions (which could probably be extended) guarantee that the message exchange works as intended.

### Issues/PRs references

~~Depends on PR #16728.~~
